crates/flox-rust-sdk/src/models/environment.rs

```diff
@@ -1,83 +1,928 @@
+    pub temp_dir: PathBuf,
+
+    /// The [EnvironmentRef] this env is created from (and validated against)
+    pub environment_ref: EnvironmentRef,
+
+    /// The transaction state
+    pub state: S,
+}
+
+#[derive(Debug, PartialEq, Eq, Copy, Clone)]
+pub struct Original;
+#[derive(Debug, PartialEq, Eq, Copy, Clone)]
+pub struct Temporary;
+
+/// A marker trait used to identify types that represent transaction states
+pub trait TransactionState: Send + Sync {}
+impl TransactionState for Original {}
+impl TransactionState for Temporary {}
+
+impl<S> PartialEq for PathEnvironment<S> {
+    fn eq(&self, other: &Self) -> bool {
+        self.path == other.path
     }
+}
 
-    pub fn system(&self) -> Cow<str> {
-        match self {
-            CommonEnvironment::Named(n) => n.system(),
-            CommonEnvironment::Project(p) => p.system(),
-        }
+impl<S: TransactionState> PathEnvironment<S> {
+    /// Makes a temporary copy of the environment so edits can be applied without modifying the original environment
+    pub async fn make_temporary(&self) -> Result<PathEnvironment<Temporary>, EnvironmentError2> {
+        copy_dir_recursively_without_permissions(&self.path, &self.temp_dir)
+            .await
+            .map_err(EnvironmentError2::MakeTemporaryEnv)?;
+        Ok(PathEnvironment {
+            path: self.temp_dir.clone(),
+            temp_dir: self.temp_dir.clone(),
+            environment_ref: self.environment_ref.clone(),
+            state: Temporary,
+        })
     }
 
-    pub fn packages(&self) -> Value {
-        todo!()
+    /// Replace the contents of this environment's `.flox` with that of another environment's `.flox`
+    pub fn replace_with(
+        &mut self,
+        replacement: PathEnvironment<Temporary>,
+    ) -> Result<(), EnvironmentError2> {
+        fs_extra::dir::move_dir(
+            &replacement.path,
+            &self.path,
+            &fs_extra::dir::CopyOptions::new()
+                .overwrite(true)
+                .content_only(true),
+        )
+        .expect("replace origin");
+        std::fs::create_dir(&replacement.path).expect("recreate temp dir");
+        Ok(())
     }
+}
 
-    pub async fn install(&self, _packages: &[FloxPackage]) {
-        todo!()
+impl PathEnvironment<Temporary> {
+    /// Updates the environment manifest with the provided contents
+    pub fn update_manifest(&mut self, contents: &impl AsRef<str>) -> Result<(), EnvironmentError2> {
+        let mut manifest_file = std::fs::OpenOptions::new()
+            .write(true)
+            .truncate(true)
+            .open(self.manifest_path())
+            .map_err(EnvironmentError2::OpenManifest)?;
+        manifest_file
+            .write_all(contents.as_ref().as_bytes())
+            .map_err(EnvironmentError2::UpdateManifest)?;
+        Ok(())
     }
+}
+
+#[async_trait]
+impl<S> Environment for PathEnvironment<S>
+where
+    S: TransactionState,
+{
```

Comment:
So i still don't see why `PathEnvironment<Temporary>` has to comply to this interface..
Yes, we want to build and use some methods we share with `PathEnvironment<Original>`, but why not implment the subset of methods as private methods for `PathEnvironment<_>` and call those in the `Environment` implementation for `PathEnvironment<Temporary>`?